File: Julia/resnet-skip-connection.jl
function compute_gradient_with_skip(gradients_F, x)
    grad = copy(x)
    for F_grad in reverse(gradients_F)
        F_mat = F_grad
        dim = size(F_mat, 2)
        grad = grad * (I + F_mat)
    end
    grad
end

function compute_gradient_without_skip(gradients_F, x)
    grad = copy(x)
    for F_grad in reverse(gradients_F)
        F_mat = F_grad
        grad = grad * F_mat
    end
    grad
end

File: R/resnet-skip-connection.R
compute_gradient_with_skip <- function(gradients_F, x) {
  grad <- x
  for (F_grad in rev(gradients_F)) {
    F_mat <- F_grad
    dim <- ncol(F_mat)
    grad <- grad %*% (diag(dim) + F_mat)
  }
  grad
}

compute_gradient_without_skip <- function(gradients_F, x) {
  grad <- x
  for (F_grad in rev(gradients_F)) {
    F_mat <- F_grad
    grad <- grad %*% F_mat
  }
  grad
}
